sisyphus/sisyphusd_test.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func NewStorageMock(t *testing.T) *StorageMock {
+	dataFile := filepath.Join(testDataDir, expectedFlakeStatsJSON)
+	raw, err := ioutil.ReadFile(dataFile)
+	if err != nil {
+		log.Fatalf("Error reading %s:%v", dataFile, err)
+	}
+	var expectedStats []FlakeStat
+	if err = json.Unmarshal([]byte(raw), &expectedStats); err != nil {
+		log.Fatalf("Failed to unmarshal test data for %s: %v", dataFile, err)
+	}
+	return &StorageMock{
+		t:             t,
+		expectedStats: expectedStats,
+	}
+}
+
+func (s *StorageMock) Store(jobName, sha string, newFlakeStat FlakeStat) error {
+	for _, expectedStat := range s.expectedStats {
+		if expectedStat.TestName == jobName && expectedStat.SHA == sha {
+			if !reflect.DeepEqual(expectedStat, newFlakeStat) {
+				s.t.Errorf("Expecting %v but got %v", expectedStat, newFlakeStat)
+			}
+			return nil
+		}
+	}
+	s.t.Errorf("No matching expectedStat found given jobName = %s, sha = %s", jobName, sha)
+	return nil
+}
+
+func TestDaemonConfig(t *testing.T) {
+	catchFlakesByRun := true
+	cfg := &Config{
+		CatchFlakesByRun: catchFlakesByRun,
+	}
+	cfgExpected := &Config{
+		CatchFlakesByRun: catchFlakesByRun,
+		PollGapDuration:  DefaultPollGapDuration,
+		NumRerun:         DefaultNumRerun,
+	}
+	sisyphusd := NewDaemon(protectedJobsMock, prowProjectMock,
+		prowZoneMock, gubernatorURLMock, gcsBucketMock, cfg)
+	if !reflect.DeepEqual(sisyphusd.GetConfig(), cfgExpected) {
+		t.Error("setting catchFlakesByRun failed")
+	}
+}
+
+func TestProwResultsMock(t *testing.T) {
+	job := "job-1"
+	prowAccessorMock := NewProwAccessorMock(gubernatorURLMock)
+	res, err := prowAccessorMock.GetProwResult(job, 10)
+	if err != nil {
+		t.Errorf("GetProwResult failed: %v", err)
+	}
+	if res.Metadata.RepoCommit != "sha-1" {
+		t.Error("RepoCommit unmatched with data in file")
+	}
+	expectedBase := 10
+	if runNo, _ := prowAccessorMock.GetLatestRun(job); runNo != expectedBase {
+		t.Errorf("Expecting frist call to GetLatestRun to return %d but got %d", expectedBase, runNo)
+	}
+	if runNo, _ := prowAccessorMock.GetLatestRun(job); runNo != expectedBase+1 {
+		t.Errorf("Expecting second call to GetLatestRun to return %d but got %d", expectedBase+1, runNo)
+	}
+	for i := 0; i < 100; i++ {
+		prowAccessorMock.GetLatestRun(job)
+	}
+	expectedMax := 15
+	if runNo, _ := prowAccessorMock.GetLatestRun(job); runNo != expectedMax {
+		t.Errorf("Expecting call to GetLatestRun to not exceed %d but got %d", expectedMax, runNo)
+	}
+}
+
+func TestRerunLogics(t *testing.T) {
+	sisyphusd := NewDaemon(protectedJobsMock, prowProjectMock,
+		prowZoneMock, gubernatorURLMock, gcsBucketMock, &Config{
+			CatchFlakesByRun: true,
+			PollGapDuration:  100 * time.Millisecond,
+		})
+	prowAccessorMock := NewProwAccessorMock(gubernatorURLMock)
+	prowAccessorMock.exitSignalToSisyphus = make(chan bool, 1)
+	sisyphusd.prowAccessor = prowAccessorMock
+	sisyphusd.SetExitSignal(prowAccessorMock.exitSignalToSisyphus)
+	sisyphusd.storage = NewStorageMock(t)
+	sisyphusd.Start()
```

Comment:
you should probably run start in a goroutine just in case it times out.

Reply:
My experience has been that a test function will return and be marked as pass once execution reaches the end, even if there are unfinished go routines. To ensure that `sisyphusd` has looped through all the mock test runs, we have to block the test function from exiting until `sisyphusd` is done, but that is equivalent to what I have here, which is to call `sisyphusd.Start()` directly. We have a 300 seconds timeout limit on testinfra presubmit so I think we should be fine.